### DISCUSSION TITLE
Compatability with Puppeteer 19?

### DISCUSSION THREAD

**Discussion by mo-rocks** [spatie/browsershot#681] [Bugs]
Puppeteer 19 downloads Chromium into a .cache folder, previously it stored it within it's own node_modules folder. For some reason Browsershot won't work anymore with this, I get the error 'Could not find expected browser locally'.

**Comment by ridxery** [comment]
I faced the same problem upgrading, tried to set Chrome path manually but none seemed to worked out so reverted it back and use older version of puppeteer right now, will try again later on

**Reply by iFlyinq** [reply]
What version have you reverted back to? I can't seem to find a version that is working at all

**Reply by ridxery** [reply]
puppeteer@13.1.1 was my old version, reverted it back everything working fine using  browsershot@3.57.2

**Comment by ethanclevenger91** [comment]
[Puppeteer 19 moved the directory the Chrome binary gets stored in](https://github.com/puppeteer/puppeteer/releases/tag/v19.0.0). The Browsershot requirements installation instructions will need to be updated.

Oddly, however, I've found that even in version 18, `.local-chromium` is not where Browsershot's requirements instructions expect it to be. 

You can either fiddle with the `chmod` command in the requirements instructions, [change Puppeteer's cache directory](https://pptr.dev/guides/configuration/#changing-the-default-cache-directory), or revert to v17. Your environment will probably dictate which routes are viable for you. I didn't want to futz with configuring the install directory, and in my Laravel Sail setup I couldn't even `cd` into the default directory where the new binary is supposed to be installed to - probably related to [this](https://github.com/puppeteer/puppeteer/issues/9192#issuecomment-1298982807). So I ended up reverting to v17:

```
curl -sL https://deb.nodesource.com/setup_14.x | sudo -E bash -
sudo apt-get install -y nodejs gconf-service libasound2 libatk1.0-0 libc6 libcairo2 libcups2 libdbus-1-3 libexpat1 libfontconfig1 libgbm1 libgcc1 libgconf-2-4 libgdk-pixbuf2.0-0 libglib2.0-0 libgtk-3-0 libnspr4 libpango-1.0-0 libpangocairo-1.0-0 libstdc++6 libx11-6 libx11-xcb1 libxcb1 libxcomposite1 libxcursor1 libxdamage1 libxext6 libxfixes3 libxi6 libxrandr2 libxrender1 libxss1 libxtst6 ca-certificates fonts-liberation libappindicator1 libnss3 lsb-release xdg-utils wget libgbm-dev libxshmfence-dev
sudo npm install --location=global --unsafe-perm puppeteer@^17
sudo chmod -R o+rx /usr/lib/node_modules/puppeteer/.local-chromium
```

Should get you up and running again.

**Reply by ethanclevenger91** [reply]
In line with this, it would be nice to see the requirements installation instructions include a fixed major version of Puppeteer and be updated only when compatibility is confirmed.

**Reply by freekmurze** [reply]
Could you send a PR for this?

**Reply by ethanclevenger91** [reply]
@freekmurze Might want to have some discussion about that first. I haven't tested it, but I'd bet that in most environments, the requirements instructions still work as-is. I think there's likely a common thread among people running into this issue. For me, it's Laravel Sail as the development environment and trying to install Puppeteer globally in Sail's PHP container. I think Docker is provisioning as a different user (or as no user?) than when you're actually using the container, so the Chrome binary doesn't end up in the expected directory or something. But I don't know enough about Docker to confirm this. I just saw someone allude to it [here](https://github.com/puppeteer/puppeteer/issues/9192#issuecomment-1298982807).

The instructions are specifically for Forge, and I don't think Forge is impacted.

A bulletproof solution would probably be to explicitly set the binary install path, though that would also require environment-specific application configuration. This may need its own section in the docs.

Or make it clear that having puppeteer be a project dependency is the preferred install method (that's an option, right?)

**Reply by ryanmortier** [reply]
Forge is definitely impacted. The instructions in the doc site do not work on my production forge server. However the instructions you posted about with v17 do work.

**Reply by RalpinMcDjango** [reply]
> Forge is definitely impacted. The instructions in the doc site do not work on my production forge server. However the instructions you posted about with v17 do work.

I second this. I've just provisioned a Ubunutu 22 server through Laravel Forge and the provided install steps do not work anymore this suggested fix has made it possible to install browser shot

**Reply by freekmurze** [reply]
Feel free to send a PR to the docs with working instructions for Forge. Thanks!

**Reply by kokoshneta** [reply]
For the record, I had the issue when trying to install on a Windows Server 2019 as well (asked about it [on Laracasts](https://laracasts.com/discuss/channels/general-discussion/browsershot-on-windows-puppeteer-cannot-find-chromium) a few months back before finding this thread). No Forge, Sail or Docker involved at all there, just a simple install directly from the command prompt on the server.

The executable is correctly downloaded and installed into `~/.cache/puppeteer`, but Puppeteer seems to be looking in `C:\Windows\system32\config\systemprofile\.cache\puppeteer` instead.

**Reply by neppoz** [reply]
The suggested solution from @ethanclevenger91 is working also for me on a 22.04 Ubuntu server. 

I would suggest to update the requirements section of the site. Me personally struggled 3 hours figuring out because it was not working. I considered these instructions -> https://spatie.be/docs/browsershot/v2/requirements

**Reply by Baspa** [reply]
Followed these steps but stil get this error message: `chmod: cannot access '/usr/lib/node_modules/puppeteer/.local-chromium': No such file or directory`. Seems my location is: `/usr/local/lib/node_modules/puppeteer/.local-chromium`.

**Comment by booni3** [comment]
I was unable to get this working using the forge install instructions. Specifically, the issue was on the last command:

```node
$ sudo chmod -R o+rx /usr/lib/node_modules/puppeteer/.local-chromium
> chmod: cannot access '/usr/lib/node_modules/puppeteer/.local-chromium': No such file or directory
```

My workaround was to install puppeteer & chromium globally. Puppeteer via NPM and chromium via apt-get.

```node
$ sudo npm install -g puppeteer
$ sudo apt-get install chromium-browser
```

Then in Browsershot, specify the chromium bin. You can use `which chromium` to find the location

```php
Browsershot::url('https://google.com')
    ->setChromePath('/snap/bin/chromium')
    ->bodyHtml()
```

I tried a few other things prior to this including:

- Installing puppeteer and chromium via npm, manually specifying all paths/bins.
- Downgrading puppeteer and global installing with npm
- The forge puppeteer instructions
- Installing puppeteer at the project level. This actually did work right away without any messing around, but I preferred to install it globally in this case.

**Reply by Baspa** [reply]
What version of Puppeteer are you using? I am still getting the following error: `Error: Cannot find module '/home/forge/<domainname>/vendor/spatie/browsershot/bin/browser.js'`

**Reply by Baspa** [reply]
For future readers with the same problems: I had to terminate Horizon to fix my issue.

**Comment by bobmurdoch** [comment]
Another way I found to make it work with Puppeteer 20.x when installed globally, if you want a more fixed Chromium install than installing it via apt-get:

1) Show dir where npm installs packages globally:
npm list -g

2) Referencing that directory, append node_modules/puppeteer/install.js to get the path to the install file. Run it:
node <path from step 1>/node_modules/puppeteer/install.js

You should see some output that it's downloading Chrome.
When complete it will give you the path to Chrome.

3) Based on the output of that command, run the permissions change
from the original instructions in this package to target the binary (what's appended to the path may change in a future version, take a look in yours to find out if needed):
sudo chmod -R o+rx <path from step 2>/chrome-linux64/chrome

4) Call browsershot with ->setChromePath() pointing to that same path.

**Reply by aeq-dev** [reply]
Hi @minutiae 
I followed your steps but still doesn't work,
Ubuntu 22
npm 9.5.1
node v18.16.0
puppeteer 20.4.0
Laravel 10

Here's my post : 
https://github.com/spatie/browsershot/discussions/736

**Reply by bobmurdoch** [reply]
The error message mentions a snap path. When I went through the steps I posted, mine didn't install via snap.
the path is more like /home/forge/.cache/puppeteer/chrome/linux-113.0.5672.63/chrome-linux64/chrome

If at some point you ran sudo apt-get install chromium-browser to install chromium (which does use snap in my setup), I'd try uninstalling that, repeating my steps, then see if the chrome binary is installed in something more like the path above.

**Reply by aeq-dev** [reply]
Thanks for reply,
In fact it's same with me, 
`/home/ubuntu/.cache/puppeteer/chrome/linux-113.0.5672.63/chrome-linux64/chrome`
So, should I set : 
`->setChromePath("/home/ubuntu/.cache/puppeteer/chrome/linux-113.0.5672.63/chrome-linux64/chrome");`

**Reply by aeq-dev** [reply]
Could you please help me for this issue :)

**Comment by colinmackinlay** [comment]
@ethanclevenger91 solution worked fine for me too on recently provisioned forge server. @freekmurze I've created a pull request #777 to update the docs

**Comment by ryanmortier** [comment]
Puppeteer is up to v21 now. Are we forever constrained to v17 or does anyone know a solution to use updated versions?

I've tried `setChromePath()` without success.

**Reply by lkmadushan** [reply]
Same problem here...

**Reply by wardhache** [reply]
For me, freezing the chromium installation to a different path worked. Know that I work using a docker container using Debian 11.
I installed chromium and placed it in a different directory (like `/tmp/chromium/`). Originally, you can find the installed files in `/usr/lib/chromium`

I've placed chromium version `120.0.6099.199` there this way. Then I've updated puppeteer to version 21.7.0.

Now, in my Dockerfile, we have the following in place:

```
ENV PUPPETEER_EXECUTABLE_PATH=/tmp/chromium/chromium
ENV PUPPETEER_SKIP_DOWNLOAD=1
...
RUN npm install --global --unsafe-perm puppeteer@21.7.0
```

It looks like I've got it working this way. Setting the skip download variable like this prevents puppeteer to download chromium and it will use the path I've provided. 

Hopefully this can help somebody

**Comment by HS-cvift** [comment]
the problem is: puppeteer is installed as root, so the browser is located in/root/ and not accessible with the forge user
the solution is simple.

skip
`curl -sL https://deb.nodesource.com/setup_14.x | sudo -E bash -` 
because the forge server had already node installed (currently v18)

install dependencies für chrome:
`sudo apt-get install -y nodejs gconf-service libasound2 libatk1.0-0 libc6 libcairo2 libcups2 libdbus-1-3 libexpat1 libfontconfig1 libgbm1 libgcc1 libgconf-2-4 libgdk-pixbuf2.0-0 libglib2.0-0 libgtk-3-0 libnspr4 libpango-1.0-0 libpangocairo-1.0-0 libstdc++6 libx11-6 libx11-xcb1 libxcb1 libxcomposite1 libxcursor1 libxdamage1 libxext6 libxfixes3 libxi6 libxrandr2 libxrender1 libxss1 libxtst6 ca-certificates fonts-liberation libappindicator1 libnss3 lsb-release xdg-utils wget libgbm-dev libxshmfence-dev` (same as documentation)

now install puppeteer as forge user
assumed: you are in the project folder, you are not using ci (vite/other node stuff)
`npm install puppeteer`

if you are using ci/vite/other node stuff, add puppeteer in your package.json and use "npm ci" in your deployment script. don't remove the node modules :)

that is all.

someone should update the documentation

**Comment by Xoshbin** [comment]
Did anyone able to make it work, for the past year that is the fourth time I'm trying to update Puppeteer, but each time I have to downgrade it again. It's working on Laravel sail, but on production on Runcloud it's giving errors.

**Comment by benholmen** [comment]
I just installed this and thought I'd share what worked for me:

```bash
curl -sL https://deb.nodesource.com/setup_22.x | sudo -E bash - # pick your nodejs version here, I chose v22 as it's the current

sudo apt-get install -y nodejs gconf-service libasound2 libatk1.0-0 libc6 libcairo2 libcups2 libdbus-1-3 libexpat1 libfontconfig1 libgbm1 libgcc1 libgconf-2-4 libgdk-pixbuf2.0-0 libglib2.0-0 libgtk-3-0 libnspr4 libpango-1.0-0 libpangocairo-1.0-0 libstdc++6 libx11-6 libx11-xcb1 libxcb1 libxcomposite1 libxcursor1 libxdamage1 libxext6 libxfixes3 libxi6 libxrandr2 libxrender1 libxss1 libxtst6 ca-certificates fonts-liberation libappindicator1 libnss3 lsb-release xdg-utils wget libgbm-dev libxshmfence-dev

sudo npm install --location=global --unsafe-perm puppeteer

npx puppeteer browsers install chrome
```

This allowed me to install the latest version of Puppeteer (v22.6.5) and Chrome (v123.0.6312.122) on a typical Forge setup.

**Reply by lamalamaMark** [reply]
Can confirm did worked for me too (with nodejs v18) on a Forge app server

**Reply by RalpinMcDjango** [reply]
I've just setup a new dev environment (WSL2 Ubuntu 22.04) with an existing codebase and your instructions were the only thing that got the browsershot library working again, thank you 👍 

I wanted to add that because I followed the current official steps in the V4 documentation before following @benholmen's instructions I also had to remove the `node_modules` folder and re-run `npm install`.

### Software Versions
- Puppeteer v23.2.2
- Node v22
- Chrome v128.0.6613.119

**Reply by christhofer** [reply]
This worked for me on Ubuntu 22.04, but the `sudo apt-get` didn't work for me on Ubuntu 24.04, I had to change some dependencies.

```diff
- sudo apt-get install -y nodejs gconf-service libasound2 libatk1.0-0 libc6 libcairo2 libcups2 libdbus-1-3 libexpat1 libfontconfig1 libgbm1 libgcc1 libgconf-2-4 libgdk-pixbuf2.0-0 libglib2.0-0 libgtk-3-0 libnspr4 libpango-1.0-0 libpangocairo-1.0-0 libstdc++6 libx11-6 libx11-xcb1 libxcb1 libxcomposite1 libxcursor1 libxdamage1 libxext6 libxfixes3 libxi6 libxrandr2 libxrender1 libxss1 libxtst6 ca-certificates fonts-liberation libappindicator1 libnss3 lsb-release xdg-utils wget libgbm-dev libxshmfence-dev

+ sudo apt-get install -y libasound2t64 libatk1.0-0t64 libc6 libcairo2 libcups2t64 libdbus-1-3 libexpat1 libfontconfig1 libgcc-s1 libgdk-pixbuf2.0-0 libglib2.0-0t64 libgtk-3-0t64 libnspr4 libpango-1.0-0 libpangocairo-1.0-0 libstdc++6 libx11-6 libx11-xcb1 libxcb1 libxcomposite1 libxcursor1 libxdamage1 libxext6 libxfixes3 libxi6 libxrandr2 libxrender1 libxss1 libxtst6 ca-certificates fonts-liberation libnss3 lsb-release xdg-utils wget libgbm-dev
```

And then I got another error `No usable sandbox!` because of [this](https://pptr.dev/troubleshooting#issues-with-apparmor-on-ubuntu).
I couldn't set up the sandbox profile correctly, and I needed a quick fix, so for now I just add this line.

```diff
  $screenshot = Browsershot::html($view->render())
+     ->noSandbox()
      ->setScreenshotType('jpeg', 90)
      ->screenshot();
```

**Comment by cornelRaiu** [comment]
If you want to use it on Vagrant with Homestead 

```
sudo apt-get install -y nodejs gconf-service libasound2 libatk1.0-0 libc6 libcairo2 libcups2 libdbus-1-3 libexpat1 libfontconfig1 libgbm1 libgcc1 libgconf-2-4 libgdk-pixbuf2.0-0 libglib2.0-0 libgtk-3-0 libnspr4 libpango-1.0-0 libpangocairo-1.0-0 libstdc++6 libx11-6 libx11-xcb1 libxcb1 libxcomposite1 libxcursor1 libxdamage1 libxext6 libxfixes3 libxi6 libxrandr2 libxrender1 libxss1 libxtst6 ca-certificates fonts-liberation libappindicator1 libnss3 lsb-release xdg-utils wget libgbm-dev libxshmfence-dev

sudo npm install --location=global --unsafe-perm puppeteer

npx puppeteer browsers install chrome
```

And then, in PHP:

```
$html = '<div>test image from section markup</div>';

Browsershot::html( $html )
    ->setChromePath( '/snap/bin/chromium' )
    ->setCustomTempPath( storage_path() )
    ->save(storage_path('app/public') . '/img.png');
```

it works with `puppeteer@22.7.1` and Chrome v124.0.6367.78 .

**Comment by liamseys** [comment]
My app runs on an ARM server in production. Does anyone have steps to get it working there? I've installed Chromium instead of Chrome.

**Reply by cornelRaiu** [reply]
My reply above gets it working on a M1 Max Mac Book Pro. Hopefully that works for you as well

**Comment by lauroguedes** [comment]
Here's my experience using a Livewire app hosted on Fly.io. I'm not sure if it's the best method, but it works for now; I might update this approach later. I've tried all the other methods mentioned, but none were successful for me.

```bash
# Install node in fly-laravel image
RUN curl -sL https://deb.nodesource.com/setup_20.x -o nodesource_setup.sh \
    && bash nodesource_setup.sh \
    && apt install -y nodejs \
    && npm install --location=global --unsafe-perm puppeteer

# Set environment for Puppeteer
ENV PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=true
ENV PUPPETEER_EXECUTABLE_PATH=/usr/bin/google-chrome

# Install dependencies for Chrome/Chromium
RUN apt-get install -y \
    wget \
    gnupg \
    ca-certificates \
    libxss1 \
    libgconf-2-4 \
    libasound2 \
    libatk1.0-0 \
    libatk-bridge2.0-0 \
    libgdk-pixbuf2.0-0 \
    libgtk-3-0 \
    libnspr4 \
    libnss3 \
    libx11-xcb1 \
    libxcomposite1 \
    libxcursor1 \
    libxdamage1 \
    libxi6 \
    libxtst6

# Install Google Chrome
RUN wget -q -O - https://dl-ssl.google.com/linux/linux_signing_key.pub | apt-key add - \
    && echo "deb [arch=amd64] http://dl.google.com/linux/chrome/deb/ stable main" >> /etc/apt/sources.list.d/google-chrome.list \
    && apt-get update -y \
    && apt-get install -y google-chrome-stable \
    && rm -rf /var/lib/apt/lists/*
```

**Reply by DiogoDuart3** [reply]
Can you share the whole Dockerfile? I tried to copy that but it didnt work :/

**Comment by DiogoDuart3** [comment]
# I managed to get it working:

Dockerfile:
```
RUN if [ -f "vite.config.js" ]; then \
        ASSET_CMD="build"; \
    else \
        ASSET_CMD="production"; \
    fi; \
    if [ -f "yarn.lock" ]; then \
        yarn install --frozen-lockfile; \
        yarn $ASSET_CMD; \
    elif [ -f "pnpm-lock.yaml" ]; then \
        corepack enable && corepack prepare pnpm@latest-8 --activate; \
        pnpm install --frozen-lockfile; \
        pnpm run $ASSET_CMD; \
    elif [ -f "package-lock.json" ]; then \
        npm ci --no-audit; \
        npm run $ASSET_CMD; \
    else \
        npm install; \
        npm run $ASSET_CMD; \
    fi;

# Environment variables for Puppeteer
ENV PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=true
ENV PUPPETEER_EXECUTABLE_PATH="/usr/bin/google-chrome"

# Install node in fly-laravel image
RUN curl -sL https://deb.nodesource.com/setup_21.x -o nodesource_setup.sh \
    && bash nodesource_setup.sh \
    && apt install -y nodejs \
    && npm install --location=global --unsafe-perm puppeteer

# From our base container created above, we
# create our final image, adding in static
# assets that we generated above
FROM base
# Environment variables for Puppeteer
ENV PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=true
ENV PUPPETEER_EXECUTABLE_PATH="/usr/bin/google-chrome"

COPY --from=node_modules_go_brrr /usr/local/bin/node /usr/local/bin/node
COPY --from=node_modules_go_brrr /app/node_modules /var/www/html/node_modules

# Install Google Chrome
RUN apt-get update && apt-get install -y wget \
    && wget -q -O /tmp/google-chrome-stable_current_amd64.deb https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb \
    && apt-get install -y /tmp/google-chrome-stable_current_amd64.deb \
    && apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/google-chrome-stable_current_amd64.deb

# Test Chrome installation
RUN google-chrome --version

# Packages like Laravel Nova may have added assets to the public directory
# or maybe some custom assets were added manually! Either way, we merge
# in the assets we generated above rather than overwrite them
COPY --from=node_modules_go_brrr /app/public /var/www/html/public-npm
RUN rsync -ar /var/www/html/public-npm/ /var/www/html/public/ \
    && rm -rf /var/www/html/public-npm \
    && chown -R www-data:www-data /var/www/html/public

# 5. Setup Entrypoint
EXPOSE 8080

ENTRYPOINT ["/entrypoint"]
```

fly.toml:
```
[env]
  APP_ENV = 'production'
  LOG_CHANNEL = 'stderr'
  LOG_LEVEL = 'info'
  LOG_STDERR_FORMATTER = 'Monolog\Formatter\JsonFormatter'
  SESSION_DRIVER = 'cookie'
  SESSION_SECURE_COOKIE = 'true'
  SSR_URL="ssr.process.laravel-react-7098.internal:13714"
  PUPPETEER_EXECUTABLE_PATH = "/usr/bin/google-chrome"
```
